C set attribute value

konstantinpersidskiy opened the pull request:
<!--
Thank you for your pull request.

If this fixes a github issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] Documentation is added or updated
- [ ] New files have a LGPL 2.1 license statement
- [ ] Tested with the following card: <!-- use `opensc-tool -n` to get the name of your card -->
	- [ ] tested PKCS#11
	- [ ] tested Windows Minidriver
	- [ ] tested macOS Tokend

Commit:
Fix data object with empty value creation

Commit:
Fix C_SetAttributeValue for CKA_VALUE for data obj

Commit:
Fix C_GetAttributeValue for attr with 0 length

Jakuje committed:
pkcs11-tool: Add examples section to the manual page

dfortinskiy1 commented on src/pkcs11/framework-pkcs15.c:
```diff
@@ -4286,6 +4299,14 @@ data_value_to_attr(CK_ATTRIBUTE_PTR attr, struct sc_pkcs15_data *data)
 		return CKR_ATTRIBUTE_VALUE_INVALID;
 
+
+	if (data->data_len == 0)
```
Блок лучше после логгера поместить.

konstantinpersidskiy replied:
Ok

dfortinskiy1 commented on src/pkcs11/framework-pkcs15.c:
```diff
@@ -4286,6 +4299,14 @@ data_value_to_attr(CK_ATTRIBUTE_PTR attr, struct sc_pkcs15_data *data)
+	if (data->data_len == 0)
+	/* value is empty */
+	{
+		attr->ulValueLen = data->data_len;
```
Здесь лучше тогда просто 0.

konstantinpersidskiy replied:
Ok

dfortinskiy1 commented on src/pkcs11/framework-pkcs15.c:
```diff
@@ -4367,7 +4388,8 @@ pkcs15_dobj_get_attribute(struct sc_pkcs11_session *session, void *object, CK_AT
 		if (rv == CKR_OK)
 			rv = data_value_to_attr(attr, data);
 		if (data) {
-			free(data->data);
```
Вообще data_value_to_attr() для других атрибутов не используется. Может и для пустого CKA_VALUE не нужно? Тогда можно будет не выделять память calloc-ом для пустого файла.

konstantinpersidskiy replied:
Ok

dfortinskiy1 commented on src/pkcs15init/pkcs15-lib.c:
```diff
@@ -3303,8 +3303,40 @@ sc_pkcs15init_change_attrib(struct sc_pkcs15_card *p15card, struct sc_profile *p
+			info->data.len = new_len;
+			u8 *nv = (u8 *) malloc (new_len * sizeof(u8));
+			memcpy(nv, new_value, new_len * sizeof(u8));
+			info->data.value = nv;
```
Разве не нужно сперва освободить старое значение info->data.value.

konstantinpersidskiy replied:
Видимо нужно

dfortinskiy1 commented on src/pkcs15init/pkcs15-lib.c:
```diff
@@ -3303,8 +3303,40 @@ sc_pkcs15init_change_attrib(struct sc_pkcs15_card *p15card, struct sc_profile *p
+			/*  set object members to represent new CKA_VALUE value,
+				new path will be written to DODF later in this function*/
+			info->data.len = new_len;
+			u8 *nv = (u8 *) malloc (new_len * sizeof(u8));
```
Я бы выделил память выше под new_data, а здесь лишь проинициализировал info.

konstantinpersidskiy replied:
Ok

frankmorgner committed:
prepare 0.18.0

marschap committed:
fix typos

Mass-typo fixing, almost exclusively in comments and text strings.

While at it also fixed a few (very few) grammar errors.

dpward committed:
Address compiler warnings when using --disable-optimization  (#1325)

* Handle errors as intended in sc_pkcs15emu_openpgp_add_data()

If a data object can be read, but it cannot be added to the PKCS#15
framework, return from this function with an error; do not continue
reading other data objects. Otherwise, do not return an error from
this function when a data object cannot be read or is empty.

Improve existing comments for clarity.

* Address other compiler warnings when using --disable-optimization

Certain variables that are not initialized when they are declared
prevent the build from completing, when --disable-optimization is
passed to ./configure.

dfortinskiy1 commented:
lgtm :thumbsup:

frankmorgner committed:
AppVeyor: let configure know that we're building with native tools

marschap committed:
openpgp: don't treat OpenPGP v3 cards special in src/libopensc/pkcs15-syn.c

As OpenPGP v3 cards are backward compatible with earlier versions,
fix an inconsitency in src/libopensc/pkcs15-syn.c
